private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,269 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Manifests are a list of paths and their hash digests, canonically ordered by
+// path in increasing lexographical order. Manifests are encoded as:
+//
+//	<hash type>:<digest>[SP][SP]<path>[LF]
+//
+// "shake256" is the only supported hash type. The digest is 64 bytes of hex
+// encoded output of SHAKE256. See golang.org/x/crypto/sha3 and FIPS 202 for
+// details on the SHAKE hash.
+package manifest
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/bufbuild/buf/private/pkg/storage"
```

Comment:
If the manifest package is meant to be small / self contained, it might be good to not pull in other dependencies if we can avoid it. Perhaps instead we could add methods to the storage package to create a manifest from a bucket?

Reply:
To be perfectly honest, I wasn't sure where to put it! It makes sense to be on its own than as a component in the storage/some-other module. There's some overlap with `Digest` and it's soon-to-exist definition in `proto/buf/alpha/module/v1alpha1/module.proto`, which likely will break that out into it's own component.

All that said, manifest construction from a storage bucket does make sense to move into the storage bucket to weaken the dependency. I'll give that a go.

Reply:
A small update: the dependency on storage comes from `NewManifestFromBucket` but it cannot live in the `storage` package either. Using `NewManifestFromBucket` implies using a storage implementation, like `storagemem`, and these packages depend on `storage`. Hence changing this dependency from `module -> storage` creates a cycle of `storage -> storagemem -> storage`.

I could create another package, perhaps as `private/pkg/manifest/storage`, that would import `manifest` and `storage`. This would break the direct dependencies on each other. But is it worth it? I don't know.